### PR TITLE
VP-4321: Fix add middleware for hagfire auditable tasks

### PR DESCRIPTION
**Pull request by Vectorfield4**
https://virtocommerce.atlassian.net/browse/VP-4321

### Proposed of changes
Describe the technical details of realization provided by you.

### Make sure these boxes are checked:
- [x] Check all the changes in github PR - files count (non of them are redundant, have meaningful changes, all are added), if target branch is correct
- [x] Check methods and variable namings - it should be self descriptive, no typos
- [x] Check you did not introduce breaking changes in API and public models/services.
- [x] Respect extensibility - https://community.virtocommerce.com/t/extensibility-basics-the-domain-model-and-persistence-layer-extension/141
- [x] Follow [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and [SOLID](https://en.wikipedia.org/wiki/SOLID) principles
- [x] For unit tests - follow Microsoft best practices: https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices
- [x] Consolidate solution dependencies in case you are using newer version, update VC module dependencies in module.manifest. Do not upgrade 3rd party packages that are shipped with the platform with the version newer than in the platform.
- [x] Check code style conventions - https://github.com/VirtoCommerce/styleguide/blob/master/csharp.md
- [x] Check PR have a concise and descriptive title, follow [git commit message rules](https://github.com/VirtoCommerce/styleguide/blob/master/gitcommits.md)

### CODE REVIEW

File: src/VirtoCommerce.Platform.Web/Middleware/Hangfire/HangfireUserContextMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+using Microsoft.AspNetCore.Http;
+using VirtoCommerce.Platform.Security;
+
+namespace VirtoCommerce.Platform.Web.Middleware.Hangfire
+{
+    public class HangfireUserContextMiddleware : IClientFilter, IServerFilter
+    {
+        private readonly IHttpContextAccessor _contextAccessor;
+
+        /// <summary>
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Web/Middleware/Hangfire/HangfireUserContextMiddleware.cs:13]: please remove this empty comment

File: src/VirtoCommerce.Platform.Web/Middleware/Hangfire/HangfireUserContextMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+using System.Threading;
+using Hangfire.Client;
+using Hangfire.Server;
+using Microsoft.AspNetCore.Http;
+using VirtoCommerce.Platform.Security;
+
+namespace VirtoCommerce.Platform.Web.Middleware.Hangfire
+{
+    public class HangfireUserContextMiddleware : IClientFilter, IServerFilter
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Web/Middleware/Hangfire/HangfireUserContextMiddleware.cs:9]: Add please a one-sentence description of what is main the responsibility of this class.

File: src/VirtoCommerce.Platform.Web/Startup.cs [hunk 3]
```diff
@@ -477,11 +477,14 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
             //Register platform settings
             app.UsePlatformSettings();
 
-            // Complete hangfire init 
+            // Complete hangfire init
             app.UseHangfire(Configuration);
 
-            app.UseModules();
+            // Add Hangfire filters/middlewares
+            var contextAccessor = app.ApplicationServices.GetRequiredService<IHttpContextAccessor>();
```
Comment (tatarincev) [inline src/VirtoCommerce.Platform.Web/Startup.cs:484]: move these lines into `ApplicationBuilderExtensions.UseHangfire`

File: src/VirtoCommerce.Platform.Web/Middleware/Hangfire/HangfireUserContextMiddleware.cs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+        /// <summary>
+        /// Ctor
+        /// </summary>
+        public HangfireUserContextMiddleware(IHttpContextAccessor contextAccessor) => _contextAccessor = contextAccessor;
+
+        /// <summary>
+        /// Check if argument contains information about log operation
+        /// </summary>
+        private readonly Func<object, bool> AuditableArgument = arg => arg is OperationLog || arg is OperationLog[];
```
Comment (yecli) [inline src/VirtoCommerce.Platform.Web/Middleware/Hangfire/HangfireUserContextMiddleware.cs]: I think it's better to fill username user for all Hangfire tasks, not just for those that are working with OperationLog.
As we have many AuditableEntities that do not have the handler which writes to operation log.
And even for those entities I think it is better to have proper user name in CreatedBy/ModifiedBy

**Review by yecli — DISMISSED** [review]

**Review by yecli — APPROVED** [review]

File: src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+        #endregion IClientFilter Members
+
+        #region IServerFilter Members
+
+        public void OnPerforming(PerformingContext filterContext)
+        {
+            var userName = filterContext.GetJobParameter<string>(USER_NAME);
+
+            Thread.SetData(Thread.GetNamedDataSlot(USER_NAME), userName);
```
Comment (yecli) [inline src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs:44]: When testing the Orders change event (e.g. change Approved and save) I've not got the proper user in [PlatformOperationLog] table.
After debugging found out that it relates to the fact that thread when this middleware.OnPerforming is executed differs from the one where DbTriggers are executed https://github.com/VirtoCommerce/vc-platform/blob/5d51ae50fddded58be04a454db8daa65e2b94518/src/VirtoCommerce.Platform.Data/Extensions/ApplicationBuilderExtensions.cs#L16, so passing data using this mechanism is not working here.

**Review by yecli — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        #endregion IClientFilter Members
+
+        #region IServerFilter Members
+
+        public void OnPerforming(PerformingContext filterContext)
+        {
+            var userName = filterContext.GetJobParameter<string>(USER_NAME);
+            if (userName != null)
+            {
+                _hangfireDataTransferService.UserName = userName;
```
Comment (yecli) [inline src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs:51]: How this solution should handle parallel hangfire task execution? 
Based on this, the same instance will be used in parallel calls - as parallelism is turned on by default in Hangfire, so Environment.ProcessorCount * 5 workers could handle jobs in parallel.
https://docs.hangfire.io/en/latest/background-processing/configuring-degree-of-parallelism.html

**Review by yecli — CHANGES_REQUESTED** [review]

File: src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs (hunk 6 above)
Comment (yecli) [reply src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs:51]: I think you need to add a test on this with multiple tasks starting the jobs under different users, to see that user name persists over one job execution.

**Comment by Vectorfield4** [comment]
It's not possible to achieve a solution in the case where we using a parallel task execution. I revert changes and set the filter to prevent setting up 'ModifyBy' if we do not receive a UserName from the thread data slot. This solution does not cover all situations and needs for workaround purposes.

File: src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs (hunk 6 above)
Comment (Vectorfield4) [reply src/VirtoCommerce.Platform.Hangfire/Middleware/HangfireUserContextMiddleware.cs]: Reverted. Check last comment https://github.com/VirtoCommerce/vc-platform/pull/2048#issuecomment-702744553

File: src/VirtoCommerce.Platform.Security/HttpContextUserResolver.cs [hunk 7]
```diff
@@ -15,7 +16,7 @@ public HttpContextUserResolver(IHttpContextAccessor httpContextAccessor)
 
         public string GetCurrentUserName()
         {
-            var result = Thread.CurrentPrincipal?.Identity?.Name ?? "unknown";
+            var result = Thread.GetData(Thread.GetNamedDataSlot(ThreadSlotNames.USER_NAME)) as string ?? "unknown";
```
Comment (yecli) [inline src/VirtoCommerce.Platform.Security/HttpContextUserResolver.cs:19]: Probably we could add a constant for "unknown" that could be used everywhere (here and in ApplicationBuilderExtensions).

**Review by yecli — APPROVED** [review]

**Pull request merged**